---
title: "Allele-specific expression analysis"
author: "Bioinformatics Platform - Institut Curie"
date: "`r Sys.Date()`"
output: html_document
fig_width: 6
fig_height: 6
---

R --no-save --no-restore -e "outPath='./'; inPath='/data/kdi_prod/.kdi/project_workspace_0/650/acl/04.00/rna_analysis/results/ascount'; min_gene_cov='10'; gene.in='/data/kdi_prod/.kdi/project_workspace_0/650/acl/04.00/rna_analysis/annotations/Mus_musculus.NCBIM37.cleaned.gtf'; known.escapees='/data/kdi_prod/.kdi/project_workspace_0/650/acl/04.00/rna_analysis/annotations/escapees_Berletch.txt'; rmarkdown::render('as_analysis_report.Rmd', output_file='${ODIR}as_analysis_report.html')"


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r, echo=FALSE}
## Load environment
require(GenomicRanges)
require(rtracklayer)
require(ggplot2)
require(reshape2)
require(RColorBrewer)
require(gridExtra)
```

```{r, echo=FALSE}
## Load variable

################################################
## Start edit here
##

#projPath <- "/data/kdi_prod/.kdi/project_workspace_0/650/acl/04.00/rna_analysis/"
#inPath <- file.path(projPath, "results/ascount/")
#outPath <- "./"
#min_gene_cov <- 10
#gene.in <- file.path(projPath, "annotations/Mus_musculus.NCBIM37.cleaned.gtf")
#known.escapees <- "/data/kdi_prod/.kdi/project_workspace_0/650/acl/04.00/rna_analysis/annotations/escapees_Berletch.txt"

#args <- commandArgs()
#print(args)

#outPath <- gsub("output_path=", "", grep("output_path", args, value = TRUE))
#inPath <- gsub("inPath=", "", grep("inPath", args, value = TRUE))
#min_gene_cov <- gsub("min_gene_cov=", "", grep("min_gene_cov", args, value = TRUE))
#gene.in <- gsub("gene.in=", "", grep("gene.in", args, value = TRUE))
#known.escapees <- gsub("known.escapees=", "", grep("known.escapees", args, value = TRUE))
##
## Stop edit here
#################################################
```

```{r, echo=FALSE}
##
## Load gene annotation
##
geneannot <- import(gene.in)
geneannot <- sortSeqlevels(geneannot)
geneannot <- sort(geneannot)
geneannot.perchr <- split(geneannot, seqnames(geneannot))
xgenes <- unique(geneannot[which(seqnames(geneannot)=="X")]$"gene_name")
autogenes <- unique(geneannot[which(seqnames(geneannot)!="X" & seqnames(geneannot) != "Y" & seqnames(geneannot) != "MT")]$"gene_name")

##
## Load data from AS mapping pipeline
##
exprs.in <- list.files(path=inPath, pattern="asratio.tsv$", full.names=TRUE)

das <- lapply(exprs.in, function(f, mcov){
  z <- read.csv(f, sep="\t", row.names=1, comment.char="#")
  z[which(rowSums(z[,1:2])>mcov),]
}, mcov=min_gene_cov)


names(das) <- basename(exprs.in)
```

##Mapping statistics

Add stats file with counts

##Per sample analysis

###Scatter plots of counts per parental genome {.tabset}

```{r, fig.height=6, fig.width=12, results="asis"}

for (i in 1:length(das)){
  cat('\n')  
  cat("####", names(das)[i], "\n") 
  x  <- das[[i]]
  xa <- x[intersect(rownames(x), autogenes),]
  p1 <- ggplot(as.data.frame(xa), aes(x=log2(genome1), y=log2(genome2))) + geom_point() + xlab("Genome 1") + ylab("Genome 2") + geom_abline(intercept=0, linetype="dashed") + annotate("text", label = "Autosomes", x=1, y = log2(max(xa$genome1)),  size = 4, colour = "darkgray")+ theme(axis.title = element_text(face="bold")) 
  xx <- x[intersect(rownames(x), xgenes),]
  p2 <- ggplot(as.data.frame(xx), aes(x=log2(genome1), y=log2(genome2), label=rownames(xx))) + geom_point() + xlab("Genome 1") + ylab("Genome 2") + geom_abline(intercept=0, linetype="dashed") + annotate("text", label = "X chromosome", x=2, y = log2(max(xa$genome1)),  size = 4, colour = "darkgray") + theme(axis.title = element_text(face="bold")) 

  #pdf(file.path(outPath, paste0(names(das)[i],"_scatter_ascounts.pdf")), width=12, height=6)
  grid.arrange(p1, p2, ncol=2)
  #dev.off()
  cat('\n') 
}
```

###Boxplots of asratio per chromosome{.tabset}

```{r, results="asis"}
for (i in 1:length(das)){
  cat('\n')  
  cat("####", names(das)[i], "\n") 
  df <- lapply(geneannot.perchr, function(gr, x){
    x[intersect(rownames(x), unique(gr$gene_name)),"ratio"]
  }, x=das[[i]])
  names(df) <- names(geneannot.perchr)
 
  d <- data.frame(x = unlist(df), 
                grp = rep(names(df), times=sapply(df, length)))
  d$grp <- factor(d$grp, levels=seqlevels(geneannot))
  
  p <- ggplot(d, aes(x = grp, y = x)) + geom_boxplot() + geom_hline(yintercept=0.5, linetype="dashed") +
    xlab("") + ylab("AS ratio") + theme(axis.title = element_text(face="bold")) 

  #pdf(file.path(outPath, paste0(names(das)[i],"_boxplot_asratio.pdf")), width=7, height=7)
  print(p)
  #dev.off()
  cat('\n') 

}

##
## Binom test ?
##

#rpv <- mclapply(as.character(splan$samples), function(i){
#  res <- apply(das[[i]][,1:2], 1, function(x){
#    if (sum(x)>0){
#      return(binom.test(x[1],x[1]+x[2],p=.5,alternative = "two.sided")$p.value)
#    }else{
#      return(NA)
#    }
#  })
#  return(res)
#})
#names(rpv) <- as.character(splan$samples)
#adjpv <- lapply(rpv, p.adjust,method="BH")
```


###Heatmaps{.tabset}

####X chromosome
```{r, fig.height=20, fig.width=4}
asHeatmap <- function(x, genes.list, show.values=FALSE, intersect=TRUE){
  stopifnot(is.list(x))

  ## Reduce the gene list to the genes detected in all samples
  if (intersect){
    genes.list <- Reduce(intersect, lapply(x, function(xx){intersect(rownames(xx), genes.list)}))
  }
  
  mat <- as.data.frame(lapply(x, function(x, k){x[k,"ratio"]}, k=genes.list))
  rownames(mat) <- genes.list
  d <- melt(as.matrix(mat))
  p <- ggplot(d, aes(y=Var1,x=Var2)) + geom_tile(aes(fill = value), colour="white") +
    scale_fill_gradientn(colours=c("red","white","steelblue"), values=c(0, 0.15, 0.85 ,1), guide=guide_legend(title="AS ratio")) +
      xlab("") + ylab("")+
        theme(axis.text.x = element_text(angle = 90, hjust = 1, size=6), axis.text.y = element_text(size=6), legend.title=element_text(size=6), legend.text=element_text(size=6))

  if (show.values){
    p <- p + geom_text(aes(label = round(value,2)), size=1.5)
  }
  
  return(p)
}

## X chromosome
## Define a list of common genes per sample

#pdf(file.path(outPath, paste0(names(das)[i],"_heatmap_X.pdf")), width=4, height=20)
asHeatmap(das, xgenes)
#dev.off()
```

####Known escapees

```{r, fig.height=7, fig.width=4}
kes <- as.vector(as.matrix(read.table(known.escapees)))
#pdf(file.path(outPath, paste0(names(das)[i],"_heatmap_known_escapees.pdf")), width=4, height=7)
asHeatmap(das, kes, show.values=TRUE)
#dev.off()
```

####Escapees

```{r, fig.height=15, fig.width=4}

##
## Heatmap of all escapees in at least one sample
## AS > .15 && AS < .85
##

allXesc <- Reduce(union, lapply(das, function(xx, k){
  xx <- xx[intersect(rownames(xx), k),]
  rownames(xx)[which(xx[,"ratio"] > .3 & xx[,"ratio"] < .7)]
}, k=xgenes))

#pdf(file.path(outPath, paste0(names(das)[i],"_heatmap_Xescapees.pdf")), width=4, height=15)
print(asHeatmap(x=das, genes.list=allXesc, show.values=TRUE, intersect=FALSE))
#dev.off()
```